File: frontend/src/types/resume.ts
// 简历相关类型定义 - 根据swagger.json定义

// ==================== 枚举类型 ====================

// 简历状态枚举
export enum ResumeStatus {
  PENDING = 'pending', // 已上传，等待解析
  PROCESSING = 'processing', // 正在解析中
  COMPLETED = 'completed', // 解析完成
  FAILED = 'failed', // 解析失败
  ARCHIVED = 'archived', // 已归档
}

// ==================== 实体类型 ====================

// 简历基础信息
export interface Resume {
  id: string;
  name: string;
  email: string;
  phone: string;
  gender?: string;
  birthday?: string;
  current_city?: string;
  highest_education?: string;
  years_experience?: number;
  status: ResumeStatus;
  resume_file_url?: string;
  parsed_at?: string;
  error_message?: string;
  uploader_id: string;
  uploader_name?: string; // 上传人姓名
  job_ids?: string[]; // 关联的岗位ID列表
  job_names?: string[]; // 关联的岗位名称列表
  job_positions?: JobApplication[]; // 关联的岗位信息
  source?: 'manual' | 'email'; // 来源：manual(手动) 或 email(邮件)
  created_at: number;
  updated_at: number;
}

// 工作经历
export interface ResumeExperience {
  id: string;
  resume_id: string;
  company: string;
  position: string;
  title: string;
  start_date: string;
  end_date: string;
  description: string;
  created_at: number;
  updated_at: number;
}

// 教育背景
export interface ResumeEducation {
  id: string;
  resume_id: string;
  school: string;
  major: string;
  degree: string;
  start_date: string;
  end_date: string;
  university_type?: 'ordinary' | '211' | '985'; // 大学类型：普通学校、211工程、985工程（已废弃，使用university_types）
  university_types?: Array<
    'ordinary' | '211' | '985' | 'double_first_class' | 'qs_top100'
  >; // 大学类型列表
  created_at: number;
  updated_at: number;
}

// 技能信息
export interface ResumeSkill {
  id: string;
  resume_id: string;
  skill_name: string;
  level: string;
  description: string;
  created_at: number;
  updated_at: number;
}

// 项目经历 - 根据swagger文档domain.ResumeProject定义
export interface ResumeProject {
  id: string;
  resume_id: string;
  name: string; // 项目名称
  project_name?: string; // 兼容旧字段
  description: string; // 项目描述
  start_date: string; // 开始日期
  end_date: string; // 结束日期
  achievements?: string; // 项目成就
  company?: string; // 所属公司
  project_type?: string; // 项目类型
  project_url?: string; // 项目链接
  responsibilities?: string; // 项目职责
  role?: string; // 担任角色
  technologies?: string; // 技术栈
  tech_stack?: string; // 兼容旧字段
  created_at: number;
  updated_at: number;
}

// 简历岗位申请关联信息 - 根据backend domain.JobApplication定义
export interface JobApplication {
  id: string;
  resume_id: string;
  job_position_id: string;
  job_title: string; // 岗位标题
  job_department: string; // 岗位部门
  status: string; // 申请状态
  source: string; // 申请来源
  notes: string; // 备注信息
  applied_at?: string; // 申请时间
  created_at: number;
  updated_at: number;
}

// 简历日志
export interface ResumeLog {
  id: string;
  resume_id: string;
  action: string;
  message: string;
  created_at: number;
  updated_at: number;
}

// 简历详细信息（包含关联数据）
export interface ResumeDetail extends Resume {
  experiences: ResumeExperience[];
  educations: ResumeEducation[];
  skills: ResumeSkill[];
  projects: ResumeProject[];
  logs: ResumeLog[];
}

// ==================== API 请求/响应类型 ====================

// 简历列表查询参数 - 根据swagger定义
export interface ResumeListParams {
  page?: number;
  size?: number;
  next_token?: string;
  position?: string;
  status?: string;
  keywords?: string;
  job_position_id?: string; // 单个岗位ID筛选（后端只支持单个）
  job_position_ids?: string[]; // 保留兼容性，但实际使用job_position_id
}

// 简历搜索参数
export interface ResumeSearchParams extends ResumeListParams {
  keywords: string;
}

// 简历列表响应 - 根据swagger ListResumeResp定义
export interface ResumeListResponse {
  resumes: Resume[];
  total_count: number;
  has_next_page: boolean;
  next_token?: string;
}

// 简历搜索响应 - 根据swagger SearchResumeResp定义
export interface ResumeSearchResponse {
  resumes: Resume[];
  total_count: number;
  has_next_page: boolean;
  next_token?: string;
}

// 简历更新参数 - 根据swagger UpdateResumeReq定义
export interface ResumeUpdateParams {
  name?: string;
  email?: string;
  phone?: string;
  gender?: string;
  birthday?: string;
  current_city?: string;
  highest_education?: string;
  years_experience?: number;
  job_ids?: string[]; // 关联的岗位ID列表（废弃，保留兼容）
  job_position_ids?: string[]; // 关联的岗位ID列表（新字段）
  experiences?: UpdateResumeExperience[];
  educations?: UpdateResumeEducation[];
  skills?: UpdateResumeSkill[];
  projects?: UpdateResumeProject[];
}

// 更新工作经历参数
export interface UpdateResumeExperience {
  id?: string; // 更新时必填
  action: 'create' | 'update' | 'delete'; // 操作类型
  company?: string;
  position?: string;
  title?: string;
  start_date?: string;
  end_date?: string;
  description?: string;
}

// 更新教育背景参数
export interface UpdateResumeEducation {
  id?: string; // 更新时必填
  action: 'create' | 'update' | 'delete'; // 操作类型
  school?: string;
  major?: string;
  degree?: string;
  start_date?: string;
  end_date?: string;
  university_type?: 'ordinary' | '211' | '985'; // 大学类型：普通学校、211工程、985工程
}

// 更新技能参数
export interface UpdateResumeSkill {
  id?: string; // 更新时必填
  action: 'create' | 'update' | 'delete'; // 操作类型
  skill_name?: string;
  level?: string;
  description?: string;
}

// 更新项目经历参数 - 根据swagger文档扩展
export interface UpdateResumeProject {
  id?: string; // 更新时必填
  action: 'create' | 'update' | 'delete'; // 操作类型
  name?: string; // 项目名称
  project_name?: string; // 兼容旧字段
  description?: string; // 项目描述
  start_date?: string; // 开始日期
  end_date?: string; // 结束日期
  achievements?: string; // 项目成就
  company?: string; // 所属公司
  project_type?: string; // 项目类型
  project_url?: string; // 项目链接
  responsibilities?: string; // 项目职责
  role?: string; // 担任角色
  technologies?: string; // 技术栈
  tech_stack?: string; // 兼容旧字段
}

// 简历解析进度 - 根据swagger ResumeParseProgress定义
export interface ResumeParseProgress {
  resume_id: string;
  status: ResumeStatus;
  progress: number; // 0-100
  message: string;
  error_message?: string;
  started_at: string;
  completed_at?: string;
}

// ==================== 业务逻辑类型 ====================

// 简历状态筛选类型
export type ResumeStatusFilter = 'all' | ResumeStatus;

// 简历筛选条件
export interface ResumeFilters {
  position?: string;
  status: ResumeStatusFilter;
  keywords: string;
}

// 分页信息
export interface PaginationInfo {
  current: number;
  total: number;
  pageSize: number;
  hasNextPage: boolean;
  nextToken?: string;
}

File: backend/internal/resume/repo/resume.go
package repo

import (
	"context"
	"fmt"
	"strings"

	"entgo.io/ent/dialect/sql"
	"github.com/google/uuid"

	"github.com/chaitin/WhaleHire/backend/db"
	"github.com/chaitin/WhaleHire/backend/db/resume"
	"github.com/chaitin/WhaleHire/backend/db/resumedocumentparse"
	"github.com/chaitin/WhaleHire/backend/db/resumeeducation"
	"github.com/chaitin/WhaleHire/backend/db/resumeexperience"
	"github.com/chaitin/WhaleHire/backend/db/resumelog"
	"github.com/chaitin/WhaleHire/backend/db/resumeskill"
	"github.com/chaitin/WhaleHire/backend/domain"
	"github.com/chaitin/WhaleHire/backend/pkg/entx"
)

type ResumeRepo struct {
	db *db.Client
}

func NewResumeRepo(db *db.Client) domain.ResumeRepo {
	return &ResumeRepo{db: db}
}

// Create 创建简历
func (r *ResumeRepo) Create(ctx context.Context, resume *db.Resume) (*db.Resume, error) {
	creator := r.db.Resume.Create().
		SetUploaderID(resume.UploaderID).
		SetName(resume.Name).
		SetGender(resume.Gender).
		SetEmail(resume.Email).
		SetPhone(resume.Phone).
		SetCurrentCity(resume.CurrentCity).
		SetHighestEducation(resume.HighestEducation).
		SetYearsExperience(resume.YearsExperience).
		SetResumeFileURL(resume.ResumeFileURL).
		SetStatus(resume.Status)

	// 处理可空字段
	if !resume.Birthday.IsZero() {
		creator = creator.SetBirthday(resume.Birthday)
	}
	if resume.ErrorMessage != "" {
		creator = creator.SetErrorMessage(resume.ErrorMessage)
	}
	if !resume.ParsedAt.IsZero() {
		creator = creator.SetParsedAt(resume.ParsedAt)
	}

	return creator.Save(ctx)
}

// GetByID 根据ID获取简历
func (r *ResumeRepo) GetByID(ctx context.Context, id string) (*db.Resume, error) {
	resumeID, err := uuid.Parse(id)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.Resume.Query().
		Where(resume.ID(resumeID)).
		WithUser().
		WithEducations().
		WithExperiences().
		WithSkills().
		WithLogs().
		Only(ctx)
}

// Update 更新简历
func (r *ResumeRepo) Update(ctx context.Context, id string, fn func(*db.Tx, *db.Resume, *db.ResumeUpdateOne) error) (*db.Resume, error) {
	resumeID, err := uuid.Parse(id)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	var result *db.Resume
	err = entx.WithTx(ctx, r.db, func(tx *db.Tx) error {
		resume, err := tx.Resume.Query().
			Where(resume.ID(resumeID)).
			Only(ctx)
		if err != nil {
			return err
		}

		updater := tx.Resume.UpdateOneID(resumeID)
		if err := fn(tx, resume, updater); err != nil {
			return err
		}

		result, err = updater.Save(ctx)
		return err
	})

	return result, err
}

// Delete 删除简历
func (r *ResumeRepo) Delete(ctx context.Context, id string) error {
	resumeID, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid resume ID: %w", err)
	}

	// 使用事务确保级联删除的原子性
	return entx.WithTx(ctx, r.db, func(tx *db.Tx) error {
		// 删除简历教育经历
		if _, err := tx.ResumeEducation.Delete().Where(
			resumeeducation.ResumeID(resumeID),
		).Exec(ctx); err != nil {
			return fmt.Errorf("failed to delete resume educations: %w", err)
		}

		// 删除简历工作经历
		if _, err := tx.ResumeExperience.Delete().Where(
			resumeexperience.ResumeID(resumeID),
		).Exec(ctx); err != nil {
			return fmt.Errorf("failed to delete resume experiences: %w", err)
		}

		// 删除简历技能
		if _, err := tx.ResumeSkill.Delete().Where(
			resumeskill.ResumeID(resumeID),
		).Exec(ctx); err != nil {
			return fmt.Errorf("failed to delete resume skills: %w", err)
		}

		// 删除简历日志
		if _, err := tx.ResumeLog.Delete().Where(
			resumelog.ResumeID(resumeID),
		).Exec(ctx); err != nil {
			return fmt.Errorf("failed to delete resume logs: %w", err)
		}

		// 删除简历文档解析记录
		if _, err := tx.ResumeDocumentParse.Delete().Where(
			resumedocumentparse.ResumeID(resumeID),
		).Exec(ctx); err != nil {
			return fmt.Errorf("failed to delete resume document parses: %w", err)
		}

		// 最后删除简历主记录
		if err := tx.Resume.DeleteOneID(resumeID).Exec(ctx); err != nil {
			return fmt.Errorf("failed to delete resume: %w", err)
		}

		return nil
	})
}

// List 获取简历列表
func (r *ResumeRepo) List(ctx context.Context, req *domain.ListResumeReq) ([]*db.Resume, *db.PageInfo, error) {
	query := r.db.Resume.Query()

	// 应用过滤条件
	if req.UploaderID != nil {
		uploaderID, err := uuid.Parse(*req.UploaderID)
		if err != nil {
			return nil, nil, fmt.Errorf("invalid uploader ID: %w", err)
		}
		query = query.Where(resume.UploaderID(uploaderID))
	}

	if req.Status != nil {
		query = query.Where(resume.Status(string(*req.Status)))
	}

	if req.Search != nil && *req.Search != "" {
		query = query.Where(
			resume.Or(
				resume.NameContains(*req.Search),
				resume.EmailContains(*req.Search),
				resume.PhoneContains(*req.Search),
				resume.CurrentCityContains(*req.Search),
			),
		)
	}

	if req.DateFrom != nil {
		query = query.Where(resume.CreatedAtGTE(*req.DateFrom))
	}

	if req.DateTo != nil {
		query = query.Where(resume.CreatedAtLTE(*req.DateTo))
	}

	// 排序
	query = query.Order(resume.ByCreatedAt(sql.OrderDesc()))

	// 分页
	return query.Page(ctx, req.Pagination.Page, req.Pagination.Size)
}

// Search 搜索简历
func (r *ResumeRepo) Search(ctx context.Context, req *domain.SearchResumeReq) ([]*db.Resume, *db.PageInfo, error) {
	query := r.db.Resume.Query()

	// 用户ID过滤
	if req.UploaderID != nil {
		uploaderID, err := uuid.Parse(*req.UploaderID)
		if err != nil {
			return nil, nil, fmt.Errorf("invalid uploader ID: %w", err)
		}
		query = query.Where(resume.UploaderID(uploaderID))
	}

	// 关键词搜索
	if req.Keywords != nil && *req.Keywords != "" {
		keywords := strings.Fields(*req.Keywords)
		var predicates []func(*sql.Selector)

		for _, keyword := range keywords {
			keyword = "%" + keyword + "%"
			predicates = append(predicates, func(s *sql.Selector) {
				s.Where(sql.Or(
					sql.Like(resume.FieldName, keyword),
					sql.Like(resume.FieldEmail, keyword),
					sql.Like(resume.FieldPhone, keyword),
					sql.Like(resume.FieldCurrentCity, keyword),
				))
			})
		}

		for _, predicate := range predicates {
			query = query.Where(predicate)
		}
	}

	// 技能搜索
	if len(req.Skills) > 0 {
		query = query.Where(
			resume.HasSkillsWith(
				resumeskill.SkillNameIn(req.Skills...),
			),
		)
	}

	// 教育背景
	if req.Education != nil && *req.Education != "" {
		query = query.Where(resume.HighestEducationContains(*req.Education))
	}

	// 工作年限
	if req.YearsExperience != nil {
		query = query.Where(resume.YearsExperienceGTE(*req.YearsExperience))
	}

	// 城市
	if req.CurrentCity != nil && *req.CurrentCity != "" {
		query = query.Where(resume.CurrentCityContains(*req.CurrentCity))
	}

	// 性别
	if req.Gender != nil && *req.Gender != "" {
		query = query.Where(resume.Gender(*req.Gender))
	}

	// 年龄范围（基于生日计算）
	if req.AgeFrom != nil || req.AgeTo != nil {
		// TODO: 实现基于生日的年龄范围查询
		// 这里需要根据生日计算年龄，暂时跳过复杂的SQL计算
		// 可以在业务层进行过滤
		_ = req.AgeFrom // 避免未使用变量警告
		_ = req.AgeTo   // 避免未使用变量警告
	}

	// 排序
	query = query.Order(resume.ByCreatedAt(sql.OrderDesc()))

	// 分页
	return query.Page(ctx, req.Pagination.Page, req.Pagination.Size)
}

// CreateEducation 创建教育经历
func (r *ResumeRepo) CreateEducation(ctx context.Context, education *db.ResumeEducation) (*db.ResumeEducation, error) {
	creator := r.db.ResumeEducation.Create().
		SetResumeID(education.ResumeID).
		SetSchool(education.School).
		SetDegree(education.Degree).
		SetMajor(education.Major)

	if !education.StartDate.IsZero() {
		creator = creator.SetStartDate(education.StartDate)
	}
	if !education.EndDate.IsZero() {
		creator = creator.SetEndDate(education.EndDate)
	}

	return creator.Save(ctx)
}

// CreateExperience 创建工作经历
func (r *ResumeRepo) CreateExperience(ctx context.Context, experience *db.ResumeExperience) (*db.ResumeExperience, error) {
	creator := r.db.ResumeExperience.Create().
		SetResumeID(experience.ResumeID).
		SetCompany(experience.Company).
		SetPosition(experience.Position).
		SetTitle(experience.Title).
		SetDescription(experience.Description)

	if !experience.StartDate.IsZero() {
		creator = creator.SetStartDate(experience.StartDate)
	}
	if !experience.EndDate.IsZero() {
		creator = creator.SetEndDate(experience.EndDate)
	}

	return creator.Save(ctx)
}

// CreateSkill 创建技能
func (r *ResumeRepo) CreateSkill(ctx context.Context, skill *db.ResumeSkill) (*db.ResumeSkill, error) {
	return r.db.ResumeSkill.Create().
		SetResumeID(skill.ResumeID).
		SetSkillName(skill.SkillName).
		SetLevel(skill.Level).
		SetDescription(skill.Description).
		Save(ctx)
}

// CreateLog 创建操作日志
func (r *ResumeRepo) CreateLog(ctx context.Context, log *db.ResumeLog) (*db.ResumeLog, error) {
	return r.db.ResumeLog.Create().
		SetResumeID(log.ResumeID).
		SetAction(log.Action).
		SetMessage(log.Message).
		Save(ctx)
}

// UpdateStatus 更新简历状态
func (r *ResumeRepo) UpdateStatus(ctx context.Context, id string, status domain.ResumeStatus) error {
	resumeID, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.Resume.UpdateOneID(resumeID).
		SetStatus(string(status)).
		Exec(ctx)
}

// UpdateErrorMessage 更新错误信息
func (r *ResumeRepo) UpdateErrorMessage(ctx context.Context, id string, errorMsg string) error {
	resumeID, err := uuid.Parse(id)
	if err != nil {
		return fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.Resume.UpdateOneID(resumeID).
		SetErrorMessage(errorMsg).
		Exec(ctx)
}

// GetEducationsByResumeID 根据简历ID获取教育经历
func (r *ResumeRepo) GetEducationsByResumeID(ctx context.Context, resumeID string) ([]*db.ResumeEducation, error) {
	resumeUUID, err := uuid.Parse(resumeID)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.ResumeEducation.Query().
		Where(resumeeducation.ResumeIDEQ(resumeUUID)).
		Order(resumeeducation.ByStartDate(sql.OrderDesc())).
		All(ctx)
}

// GetExperiencesByResumeID 根据简历ID获取工作经验
func (r *ResumeRepo) GetExperiencesByResumeID(ctx context.Context, resumeID string) ([]*db.ResumeExperience, error) {
	resumeUUID, err := uuid.Parse(resumeID)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.ResumeExperience.Query().
		Where(resumeexperience.ResumeIDEQ(resumeUUID)).
		Order(resumeexperience.ByStartDate(sql.OrderDesc())).
		All(ctx)
}

// GetSkillsByResumeID 根据简历ID获取技能
func (r *ResumeRepo) GetSkillsByResumeID(ctx context.Context, resumeID string) ([]*db.ResumeSkill, error) {
	resumeUUID, err := uuid.Parse(resumeID)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.ResumeSkill.Query().
		Where(resumeskill.ResumeIDEQ(resumeUUID)).
		Order(resumeskill.ByCreatedAt(sql.OrderDesc())).
		All(ctx)
}

// GetLogsByResumeID 根据简历ID获取操作日志
func (r *ResumeRepo) GetLogsByResumeID(ctx context.Context, resumeID string) ([]*db.ResumeLog, error) {
	resumeUUID, err := uuid.Parse(resumeID)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.ResumeLog.Query().
		Where(resumelog.ResumeIDEQ(resumeUUID)).
		Order(resumelog.ByCreatedAt(sql.OrderDesc())).
		All(ctx)
}

// CreateDocumentParse 创建文档解析记录
func (r *ResumeRepo) CreateDocumentParse(ctx context.Context, documentParse *db.ResumeDocumentParse) (*db.ResumeDocumentParse, error) {
	creator := r.db.ResumeDocumentParse.Create().
		SetResumeID(documentParse.ResumeID).
		SetFileID(documentParse.FileID).
		SetContent(documentParse.Content).
		SetFileType(documentParse.FileType).
		SetFilename(documentParse.Filename).
		SetTitle(documentParse.Title).
		SetStatus(documentParse.Status)

	if !documentParse.UploadAt.IsZero() {
		creator = creator.SetUploadAt(documentParse.UploadAt)
	}
	if documentParse.ErrorMessage != "" {
		creator = creator.SetErrorMessage(documentParse.ErrorMessage)
	}

	return creator.Save(ctx)
}

// GetDocumentParseByResumeID 根据简历ID获取文档解析记录
func (r *ResumeRepo) GetDocumentParseByResumeID(ctx context.Context, resumeID string) (*db.ResumeDocumentParse, error) {
	resumeUUID, err := uuid.Parse(resumeID)
	if err != nil {
		return nil, fmt.Errorf("invalid resume ID: %w", err)
	}

	return r.db.ResumeDocumentParse.Query().
		Where(resumedocumentparse.ResumeIDEQ(resumeUUID)).
		Order(resumedocumentparse.ByCreatedAt(sql.OrderDesc())).
		First(ctx)
}
